### PR TITLE
impl(otel): support non-string resource attributes

### PR DESCRIPTION
**Pull request by dbolduc**
Related to #11775 

In our Cloud Trace and Cloud Monitoring exporters, we copy over resource attributes. Our GCP Resource Detector only uses strings for the value of these attributes. But in the off-chance that an application is using a different resource detector, handle these additional attribute types.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13902)

### CODE REVIEW

File: google/cloud/opentelemetry/internal/monitored_resource_test.cc [hunk 1]
```diff
@@ -27,6 +27,46 @@ namespace sc = opentelemetry::sdk::resource::SemanticConventions;
...
+TEST(AsString, Simple) {
+  struct TestCase {
+    opentelemetry::sdk::common::OwnedAttributeValue value;
+    std::string result;
+  };
+  std::vector<TestCase> cases = {
+      {true, "true"},       {false, "false"},
+      {int32_t{1}, "1"},    {uint32_t{2}, "2"},
+      {int64_t{3}, "3"},    {uint64_t{4}, "4"},
+      {double{5.6}, "5.6"}, {std::string{"value"}, "value"},
```
Comment (devbww) [inline google/cloud/opentelemetry/internal/monitored_resource_test.cc:39]: Perhaps we should quote string attributes to distinguish the output of `true` and `"true"`.

Comment (dbolduc) [reply google/cloud/opentelemetry/internal/monitored_resource_test.cc:39]: I considered that, but I don't think preserving the type is a feature. The majority of these are going to be strings. And in other places we expect the resulting string to not be quoted.

File: google/cloud/opentelemetry/internal/monitored_resource.cc [hunk 2]
```diff
@@ -29,15 +32,25 @@ namespace {
...
 namespace sc = opentelemetry::sdk::resource::SemanticConventions;
 
-std::string AsString(
-    opentelemetry::sdk::common::OwnedAttributeValue attribute) {
-  if (absl::holds_alternative<std::string>(attribute)) {
-    return absl::get<std::string>(attribute);
+struct AsStringVisitor {
+  template <typename T>
+  std::string operator()(std::vector<T> const& v) const {
+    return absl::StrCat("[", absl::StrJoin(v, ", "), "]");
```
Comment (devbww) [inline google/cloud/opentelemetry/internal/monitored_resource.cc:38]: Consider whether the vector versions should call through to the scalar versions.

File: google/cloud/opentelemetry/internal/monitored_resource_test.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/opentelemetry/internal/monitored_resource_test.cc:39]: OK.

**Review by devbww — APPROVED** [review]

**Pull request merged**